### ISSUE TITLE
SCSS property shorthand issues

### ISSUE THREAD

**Issue by lydell** [stylelint/stylelint#2177]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

A bug where stylelint fails to parse valid SCSS code. Also, a bug where a "declaration" is mistaken as a rule.

> Which rule, if any, is this issue related to?

- `rule-nested-empty-line-before`
- `declaration-empty-line-before`
- probably any rule that needs to parse selectors

> What CSS is needed to reproduce this issue?

Note: <strong>S</strong>CSS.

```scss
.foo {
  color: black;
  background: {
    size: cover;
    position: center;
  }
  padding: 10px;
}
```

> What stylelint configuration is needed to reproduce this issue?

```js
module.exports = {
  extends: 'stylelint-config-standard',
}
```

> Which version of stylelint are you using?

`7.6.0`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

```json
"scripts": {
  "test": "stylelint test.scss --syntax scss"
}
```

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

Yes, it's related to SCSS property shorthands (not sure if that's exactly what they're called).

```scss
background: {
  size: cover;
  position: center;
}
```

> What did you expect to happen?

No warnings to be flagged.

> What actually happened (e.g. what warnings or errors you are getting)?

The following warnings were flagged:

```
test.scss
 3:3  ✖  Expected empty line before nested rule   rule-nested-empty-line-before
 3:3     Cannot parse selector                                                 
 3:3     Cannot parse selector                                                 
 3:3     Cannot parse selector                                                 
 3:3     Cannot parse selector                                                 
 3:3     Cannot parse selector                                                 
 3:3     Cannot parse selector                                                 
 3:3     Cannot parse selector                                                 
 7:3  ✖  Expected empty line before declaration   declaration-empty-line-before
```

Workaround: Don't use SCSS property shorthands.

---

Easiest way to reproduce:

1. Clone this gist: https://gist.github.com/lydell/64f3f4eb41b162c7443def5d68b27708
2. `npm install`
3. `npm test`

I'm willing to work on fixing this after some discussion on how to attack this problem.

**Comment by evilebottnawi** [comment]
@lydell we fixed some regression errors after `7.6.0` release with non-standard syntax, can you test you example on `master` branch?

**Comment by lydell** [comment]
This is the output when using commit 8c3e0d6c646b912d69285412498e372168bc4815:

```
test.scss
 7:3  ✖  Expected empty line before declaration   declaration-empty-line-before
```

**Comment by Airkro** [comment]
i get this error too, i think Sass nested properties may not be supported

**Comment by evilebottnawi** [comment]
@lydell reproduce, accepted

**Comment by jeddy3** [comment]
> Sass nested properties may not be supported

That is correct.

> This is the output when using commit 8c3e0d6:

I think the rule is behaving as expected.

For the following config (from within `stylelint-config-standard`):

```json
{
"declaration-empty-line-before": [ "always", {
      except: [
        "after-declaration",
        "first-nested",
      ],
      ignore: [
        "after-comment",
        "inside-single-line-block",
      ],
    } ],
}
```

The following code:

```css
.foo {
  color: black;
  background: {
    size: cover;
    position: center;
  }
  padding: 10px;
}
```

Produces this warning:

```test.scss
 7:3  ✖  Expected empty line before declaration   declaration-empty-line-before
```

There should always be an empty line before a declaration, unless it comes after a declaration or is the first nested thing. And on line 7 there's a declaration _coming after an unknown construct_ that doesn't have an empty line before it. 

@lydell I think you [have a couple of options](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/css-processors.md#css-processors):

1. Don't use this construct.
2. Preprocess this construct before linting the code, perhaps using [`postcss-nested-props`](https://github.com/jedmao/postcss-nested-props) and [`postcss-scss`](https://github.com/postcss/postcss-scss).
3. See how much mileage you get from the `"never"` option.
4. Add a plugin to [`stylelint-scss`](https://github.com/kristerkari/stylelint-scss) that understands this construct.

You'll need to weigh-up advantages and costs of using this construct. For example, I can think of handful of other core rules this construct renders useless e.g. `property-no-unknown`, the `declaration-property-*-*list` rules.

**Comment by lydell** [comment]
I’m going to:

1. Update stylelint to some commit newer than 7.6.0 (awaiting the next release).
2. Add newlines to please the `declaration-empty-line-before` rule. Doesn’t really matter to me.
3. Consider whether this shorthand syntax is worth it, considering that it might break, or remove the benefit from, other rules (as mentioned by @jeddy3 ).

Thanks for the help everyone!

**Comment by Airkro** [comment]
@jeddy3 

I don't get this error any more when I using vscode-stylelint 0.21.1
It 's already fixed or not？

**Comment by evilebottnawi** [comment]
@Airkro yes
> Fixed: selector-max-compound-selectors ignores SCSS nested declarations (#2102).
